056-zk-kraft-migration.md

```diff
@@ -0,0 +1,254 @@
+The FSM moves to a `WaitControllersInMigration` state where the operator expects the user to create the KRaft controllers pool.
+The user creates the `KafkaNodePool` custom resource containing the definition of the KRaft controllers which are going to replace the ZooKeeper nodes.
+The operator creates and configures the controllers accordingly with the ZooKeeper migration flag and the connection to the ZooKeeper ensemble. 
+
+```shell
+# Enable the migration
+zookeeper.metadata.migration.enable=true
+
+# ZooKeeper client configuration
+zookeeper.connect=<connection-to-zookeeper>
+```
+
+The operator has to wait for the KRaft controllers being ready, the quorum to be formed and leader elected, then the migration process moves to the state of waiting for brokers to register.
+
+The FSM moves to a `WaitBrokersInMigration` state where the operator expects the user to ask for running the actual metadata migration.
+It means the operator has to enable the ZooKeeper migration on the brokers pool in order to actually start the migration (Phase 2).
+
+From this state it's still possible to not move forward with the migration.
+The user can delete the `KafkaNodePool` custom resource containing the definition of the KRaft controllers and remove the `strimzi.io/kraft-migration` annotation from the `Kafka` custom resource.
+The FSM moves back to the `NoMigration` state (or null) actually exiting from the migration process and the operator upgrades `Kafka.status.migrationState` field accordingly.
+There is nothing to do with the brokers which are still configured to work with the ZooKeeper ensemble only.
+
+#### Phase 2: Enabling the migration on the brokers
+
+In the `WaitBrokersInMigration` state, the operator expects the user to ask for running the actual metadata migration.
+In order to start the actual migration, the brokers need to be connected to the KRaft controllers and have the ZooKeeper migration enabled.
+The user annotates the `Kafka` custom resource running the current ZooKeeper based brokers with `strimzi.io/kraft-migration: migration`.
+This annotation drives the operator to update the brokers' configuration with the ZooKeeper migration flag enabled and with the new KRaft controllers quorum connection information as well.
+
+```shell
+# KRaft controller quorum configuration
+controller.quorum.voters=<voters>
+controller.listener.names=CONTROLLER
+
+# The controller.listener.name should also be added to listener.security.property.map
+listener.security.protocol.map=PLAINTEXT:PLAINTEXT,CONTROLLER:PLAINTEXT
+
+# Enable the migration
+zookeeper.metadata.migration.enable=true
+
+# ZooKeeper client configuration
+zookeeper.connect=<connection-to-zookeeper>
+```
+
+Finally, the operator rolls the brokers.
+After the rolling, the internal Apache Kafka migration component starts the actual metadata migration and the operator has to wait for it to be completed to move to the next phase.
```

Comment:
My understanding is that this process is async. After the last broker restarts, the active controller gets the broker registration with the migration flag enabled, and it starts the actual metadata migration from ZK. After that, it starts working in dual-write mode.

Reply:
Yeah that's how Fede explained. The reconciliation ends with brokers restarting. When they are ready again, they are connected to the KRaft controllers and because the ZooKeeper migration flag is enabled, Kafka starts the migration. The operator can get info/updates update the process querying the Kafka Agent endpoint I am envisioning for getting metrics about migration. Of course, it happens on each reconciliation cycle, until migration ends, the FSM moves to the next state (dual write) waiting for next action from the user.